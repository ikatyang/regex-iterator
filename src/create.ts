/**
 * create a generator for string matching
 */
export function* create(regex: RegExp, str: string, max_result = Infinity) {
  const cloned_regex = new RegExp(regex.source, regex.flags);

  let counter = 0;
  let result = cloned_regex.exec(str);

  while (counter < max_result && result !== null) {
    counter++;

    yield result;

    if (!regex.global) {
      break;
    }

    result = cloned_regex.exec(str);
  }
}
