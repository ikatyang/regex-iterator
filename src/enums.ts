export const enum TakeAction {
  Break = -1,
  TakeAndBreak = 0,
  TakeAndContinue = 1,
}
