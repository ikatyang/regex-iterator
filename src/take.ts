import {TakeAction} from './enums';
import {iterate} from './iterate';

/**
 * take items in the generator
 */
export const take = <T, U = T>(
    iterator: IterableIterator<T>,
    pred: (value: T) => void | boolean | TakeAction,
    fn: (value: T, index: number, values: U[]) => U = (x: T) => x as any as U,
    ): U[] => {
  const values: U[] = [];

  iterate(iterator, (value, index) => {
    const action = pred(value);

    switch (action) {

      case true:
      case TakeAction.TakeAndBreak:
      case TakeAction.TakeAndContinue:
        values.push(fn(value, index, values));
        break;

      default:
        // do nothing
        break;
    }

    return action;
  });

  return values;
};
