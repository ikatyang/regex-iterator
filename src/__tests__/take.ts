import { create } from '../create';
import { TakeAction } from '../enums';
import { take } from '../take';

const test_return_value = (return_value: any, expected_count: number) => {
  const iterator = create(/[0-9]+/g, '123 456 789');
  const values = take(iterator, () => return_value);
  expect(values.length).toBe(expected_count);
};

it('should take correctly', () => {
  test_return_value(false, 0);
  test_return_value(undefined, 0);
  test_return_value(TakeAction.Break, 0);

  test_return_value(TakeAction.TakeAndBreak, 1);

  test_return_value(true, 3);
  test_return_value(TakeAction.TakeAndContinue, 3);
});

it('should take transformed correctly', () => {
  const iterator = create(/[0-9]+/g, '123 456 789');
  const values = take(
    iterator,
    () => true,
    (value, index) => `${index}:${value[0]}`,
  );
  expect(values).toMatchSnapshot();
});
