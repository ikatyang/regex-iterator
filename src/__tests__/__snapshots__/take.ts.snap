// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take transformed correctly 1`] = `
Array [
  "0:123",
  "1:456",
  "2:789",
]
`;
